### ISSUE TITLE
SummarizedExperiment doesn't work with data.frame input

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/SummarizedExperiment#17]
I would have expected the following to work:

```r
a <- data.frame(A=1:5, B=2:6)
library(SummarizedExperiment)
SummarizedExperiment(a)
## Error in seq_len(ncol(assay)) : 
##   argument must be coercible to non-negative integer
## In addition: Warning message:
## In seq_len(ncol(assay)) : first element used of 'length.out' argument
```

This is attributable to the fact that a `data.frame` is treated as a `list` of column vectors and ends up being processed as such. The only way I get the desired result is to do:

```r
SummarizedExperiment(list(list(B=a)))
```

One could argue that I shouldn't be storing a `data.frame` as an SE assay anyway, and that would be fair enough. But in that case, I would expect a more useful error message (e.g., instructing the user to coerce to an ordinary matrix) than what we currently get, which is pretty uninterpretable.

<details>
<summary>Session information</summary>

```
R Under development (unstable) (2018-11-02 r75535)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.5 LTS

Matrix products: default
BLAS: /home/cri.camres.org/lun01/Software/R/trunk/lib/libRblas.so
LAPACK: /home/cri.camres.org/lun01/Software/R/trunk/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] SummarizedExperiment_1.13.0 DelayedArray_0.9.0         
 [3] BiocParallel_1.17.1         matrixStats_0.54.0         
 [5] Biobase_2.43.0              GenomicRanges_1.35.1       
 [7] GenomeInfoDb_1.19.1         IRanges_2.17.1             
 [9] S4Vectors_0.21.4            BiocGenerics_0.29.1        

loaded via a namespace (and not attached):
 [1] lattice_0.20-38        bitops_1.0-6           grid_3.6.0            
 [4] zlibbioc_1.29.0        XVector_0.23.0         Matrix_1.2-15         
 [7] tools_3.6.0            RCurl_1.95-4.11        compiler_3.6.0        
[10] GenomeInfoDbData_1.2.0
```
</details>

**Comment by epurdom** [comment]
I agree more useful error message would be appreciated, as I just ran into this same issue (R version 3.6.0, SummarizedExperiment 1.15.2) and it took a bit of work to figure what was the source of the error.

**Comment by Alanocallaghan** [comment]
I agree - this should be disallowed or handled in a more elegant way. The current error messages are bizarre.

If colData is included, the message is actually different:
```r
SingleCellExperiment(
  assays = list(counts = myDataFrame),
  colData = myColData
)
# Error in all_dims[, 1L] : incorrect number of dimensions
```

**Comment by hpages** [comment]
Fixed in SummarizedExperiment 1.15.9 (commit 0d74843c8923a917fecf0e2b75c95c8456da5cb3).

H.
